### DISCUSSION TITLE
PyScript should understand that `.whl` files are just `.zip` files ;-)

### DISCUSSION THREAD

**Discussion by ntoll** [pyscript/pyscript#2067] [Proposals]
A Python wheel (see spec: https://packaging.python.org/en/latest/specifications/binary-distribution-format/) is just a `.zip` file with a certain set of metadata and specific layout. Wouldn't it awesome if PyScript understood enough about the `.whl` format such that the module with the `.whl` archive was properly extracted into the Python path. This would be especially helpful, when using MicroPython with pure Python wheels. Vice versa, users (TurtleThread.com) would like to build wheels that work with both MicroPython and regular Python. cc/@marieroald

**Comment by WebReflection** [comment]
Do I understand this correctly?

> Although a specialized installer is recommended, a wheel file may be installed by simply unpacking into site-packages with the standard ‘unzip’ tool while preserving enough information to spread its contents out onto their final paths at any later time.

meaning that all we need to do is to force `.whl` as regular `.zip` files and call it a day? It might be trivial on MicroPython as we do implement the *unzip* logic ourselves but I think *Pyodide* doesn't need any special care at all there, except us accepting `.whl` too as extension, see https://pyodide.org/en/stable/usage/loading-packages.html#installing-wheels-from-arbitrary-urls

If I could get a `.whl` to test, one that works in both Pyodide and MicroPython, this looks like a no-brainer to me.

**Reply by ntoll** [reply]
This is mostly so we could use pure Python wheels with MicroPython. Definitely an experimental feature, since there's no guarantee that the content of the module contains Python that works with MicroPython's subset of the language. But tl;dr - yeah, it's just unzipping. ;-)

**Reply by WebReflection** [reply]
can I have such `.whl` example that is not made up but an actual representation of a pure Python wheel that would work in both Pyodide and MicroPython?

**Reply by ntoll** [reply]
:skull_and_crossbones: Arrr should do the trick: https://pypi.org/project/arrr/#files

**Reply by WebReflection** [reply]
@ntoll am I understanding correctly we want `.whl` files in `packages` instead of `files` ?

**Reply by WebReflection** [reply]
... or both? (but not duplicated)

**Reply by ntoll** [reply]
Good question.

In MicroPython, if the `packages` contains a URL for a `.whl` then it should unzip it instead. In Pyodide, a reference to a wheel via a URL should just be handled by micropip (as it currently is). Make sense?

**Reply by WebReflection** [reply]
MR here https://github.com/pyscript/polyscript/pull/102 ... works like a charm

**Reply by WebReflection** [reply]
MR to land latest pllyscript here https://github.com/pyscript/pyscript/pull/2084 this closes this discussion (or better, it makes it solved).
